// redux-resource-wlfe/lib/createResourceAction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createResourceAction;

var _fetchResource = require('./helpers/fetchResource');

var _fetchResource2 = _interopRequireDefault(_fetchResource);

var _omit = require('./helpers/omit');

var _omit2 = _interopRequireDefault(_omit);

var _parseUrl = require('./helpers/parseUrl');

var _parseUrl2 = _interopRequireDefault(_parseUrl);

var _assign = require('./helpers/assign');

var _assign2 = _interopRequireDefault(_assign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createResourceAction(options, sendType, successType, errorType) {
  var rawUrl = undefined,
      urlCompiler = undefined;

  if (typeof options === 'string') {
    rawUrl = options;
    options = {};
  } else {
    rawUrl = options.url;
    options = (0, _omit2.default)(options, ['url']);
  }

  if (!rawUrl) {
    throw new Error('Please provide a url for the resource');
  }

  urlCompiler = (0, _parseUrl2.default)(rawUrl);

  var resourceSendAction = function resourceSendAction() {
    return { type: sendType };
  };

  var resourceSuccessAction = function resourceSuccessAction(resource) {
    return {
      type: successType,
      payload: resource
    };
  };

  var resourceErrorAction = function resourceErrorAction(error) {
    return {
      type: errorType,
      payload: error
    };
  };

  return function (params, data) {
    return function (dispatch) {
      var url = urlCompiler(params);

      dispatch(resourceSendAction());

      return (0, _fetchResource2.default)(url, (0, _assign2.default)({}, options, { data: data })).then(function (resource) {
        return dispatch(resourceSuccessAction(resource));
      }, function (error) {
        return dispatch(resourceErrorAction(error));
      });
    };
  };
}
